inc/Engine/CriticalPath/APIClient.php

```diff
@@ -0,0 +1,230 @@
+<?php
+namespace WP_Rocket\Engine\CriticalPath;
+
+use WP_Error;
+
+class APIClient {
+
+	/**
+	 * Constant url for Critical Path API job.
+	 */
+	const API_URL = 'https://cpcss.wp-rocket.me/api/job/';
+
+	/**
+	 * Sends a generation request to the Critical Path API
+	 *
+	 * @since 3.6
+	 *
+	 * @param string $url The URL to send a CPCSS generation request for.
+	 * @return array
+	 */
+	public function send_generation_request( $url ) {
+		$response = wp_remote_post(
+			self::API_URL,
+			[
+				// This filter is documented in inc/Engine/CriticalPath/CriticalCSSGeneration.php.
+				'body' => apply_filters(
+					'rocket_cpcss_job_request',
+					[
+						'url' => $url,
+					]
+				),
+			]
+		);
+
+		return $this->prepare_response( $response, $url );
+	}
+
+	/**
+	 * Prepare the response to be returned.
+	 *
+	 * @since 3.6
+	 *
+	 * @param array|WP_Error $response The response or WP_Error on failure.
+	 * @param string         $url Url to be checked.
+	 * @return array|WP_Error
+	 */
+	private function prepare_response( $response, $url ) {
+		$response_data        = $this->get_response_data( $response );
+		$response_status_code = $this->get_response_status( $response );
+		$succeeded            = $this->get_response_success( $response_status_code, $response_data );
+
+		if ( $succeeded ) {
+			return $response_data;
+		}
+
+		$response_code    = $this->get_response_code( $response );
+		$response_message = $this->get_response_message( $response_status_code, $response_data, $url );
+
+		if ( 200 === $response_status_code ) {
+			$response_status_code = 400;
+		}
+
+		return new WP_Error(
+			$response_code,
+			$response_message,
```

Comment:
The response code and message getters can be called directly in the `new WP_Error` instead of first assigning them to vars. Why? 

- Readability and maintainability are the same by using the methods without the variables.
The method names are very clear as to what will be returned, i.e. code or message. 

- Variables not needed.
The variables are not being used anywhere else. Rather, they are only used as args for `WP_Error`.

- More performant. Why? Less memory and processing time. How? 
Let's explore how PHP internals work:
    - After `get_response_code()` runs, the code it returns is added into the memory stack.
    - Next, PHP adds the variable into the symtable (i.e. symbols table - a separate lookup in memory).
    - Next, PHP binds the variable in the symtable to the code's memory location in the stack.
    - Repeat for the message.
    - Then PHP looks up the variable in the symtable and then the code in memory to populate the first arg in `WP_Error`.
    - Repeat for the message and 2nd arg.

```suggestion
		if ( 200 === $response_status_code ) {
			$response_status_code = 400;
		}

		return new WP_Error(
			$this->get_response_code( $response ),
			$this->get_response_message( $response_status_code, $response_data, $url ),
```

Reply:
@hellofromtonya you are right and thanks for illustration, I will do that for response code but I can't do that for the message because of:-
- I need to call `get_response_message` before normalizing the value of `$response_status_code`.
- This helps me in the case of the response code is 200 but the response body is empty so here I want to pass the code 200 and inside the method `get_response_message` I can check if the code is 200 and the response body is empty get the corresponding message but the response code in this case should return 400 so I'm changing it after getting the message.